File: src/eye_analysis.cpp
#include "eye_analysis.hh" 

namespace SchematicEye {

Analysis::Analysis() {};

Analysis::~Analysis() {};


float *Analysis::IntensityAnalysis(float object_distance=1000000, float off_axis=5, 
        float pupil_size=3, float diopters=0)
{
    std::string model="navarro";
    std::string param="foo";
    int _diop = diopters;

    Eye eye = Eye();  
    eye.set_params(_diop, pupil_size, model, 20);
    eye.SchematicEye();
    eye.EyeTracer(object_distance, off_axis);
                
    eye.sys->get_tracer_params().set_default_distribution(
                                Trace::Distribution(Trace::HexaPolarDist, 250)); 
                    
    Goptical::Analysis::Spot spot(*eye.sys);  

    double maxRadius = 0.2;
    double radiusStep = 0.0005;
    int len = (maxRadius / radiusStep) - 1;
    float *outfile = new float [len];
    int i(0);
    double radius = 0.0005; // in mm.
    while ( radius <= maxRadius)
    {
        // use .get_ray_wavelen_set() to process spot for each wavelength? 
        // then wighted average?
        outfile[i] = spot.get_encircled_intensity( radius );
        i++;
        radius += radiusStep;
    }
    return outfile;
}

void Analysis::IntensityAnalysis(std::string param, float object_distance=1000000, 
                    float off_axis=5, std::string model="dubbleman", 
                    float age=20, float pupil_size=3, float diopters=0, int iter=4)
{
    int _diop = diopters;
    std::ofstream outputfile;

    // add ability to change name of Intensity
    // setup output files and headers:
    outputfile.open ("dat/Intensity.csv", std::ios::trunc);
    outputfile << "encircled_intensity" << "," 
                << "radius_mm" << "," 
                << "lens_focus_D" << "," 
                << "pupil_size_mm" << ","
                << "offaxis_deg" << ","
                << "age_y" << ","
                << "obj_distance_mm" << ","
                << "model" << ","
                << "axial_len_mm" << ","
                << "iterations" << std::endl;

    for (int i = 0; i < iter; i++)
    {    
        Eye eye = Eye();  
        eye.set_params(_diop, pupil_size, model, age);
        eye.SchematicEye();
        eye.EyeTracer(object_distance, off_axis);

        std::cout << " " <<std::endl;
        std::cout << "lens (diopters): " << _diop << "  Age (years): " 
                    << eye.age << "  pupil size (mm):" << eye.pupil_size << std::endl;

        std::cout << "off axis (deg):" << off_axis
                    << "  obj dist (mm): " << object_distance << std::endl;
                    
        std::cout << "unaccommodated defocus: " << eye.FindOpticalPower(2)
                    << "  best focus (diopter): " << eye.FindOpticalPower(1) << std::endl;
                    
        eye.sys->get_tracer_params().set_default_distribution(
                                    Trace::Distribution(Trace::HexaPolarDist, 250)); 
                        
        Goptical::Analysis::Spot spot(*eye.sys);  

        float eye_len, pup_s;
        eye_len = eye.ReturnAxialLength();
        pup_s = eye.ReturnPupilSize();
        double radius = 0.0005; // in mm.
        while ( radius <= 0.2)
        {
            // use .get_ray_wavelen_set() to process spot for each wavelength? 
            // then wighted average?
            outputfile << spot.get_encircled_intensity( radius ) 
                << "," << radius
                << "," << _diop 
                << "," << pup_s
                << "," << off_axis 
                << "," << age 
                << "," << object_distance 
                << "," << model 
                << "," << eye_len
                << "," << iter << std::endl;
            
            radius += 0.0005;
        }
        // increment the chosen variable.
        if (param == "age") { age += 2; }
        if (param == "pupil") { pupil_size++; }
        if (param == "focus") { _diop += 2; }
        if (param == "angle") { off_axis += 5; }
        if (param == "distance") { object_distance = pow(10, 4 - i); }
    }

}


void Analysis::EyePlots(int best_focus = 1, float object_distance=100000, 
                    float off_axis=5, std::string model="dubbleman", 
                    float age=20, float pupil_size=3, float diopters=0)
{    

    switch (best_focus) 
    {
        case 1:
        {
            Eye eye = Eye(); 
            eye.set_params(diopters, pupil_size, model, age);
            eye.SchematicEye();
            eye.EyeTracer(object_distance, off_axis);

            Goptical::Io::RendererSvg renderer("img/eye.svg", 1200,1200);
            renderer.set_margin_ratio(0.35, 0.25, 0.1, 0.1);

            // layout plot
            renderer.set_page_layout(1,2);
            
            // draw 2d system layout: 
            eye.sys->draw_2d_fit(renderer);
            eye.sys->draw_2d(renderer);

            // trace and draw rays from rays source
            eye.sys->enable_single<Sys::Source>(*eye.source_rays);
            eye.tracer->get_trace_result().set_generated_save_state(*eye.source_rays);
            
            renderer.set_page(0);
            eye.tracer->trace();
            eye.tracer->get_trace_result().draw_2d(renderer, true); //, eye.image);
            
            // longitudinal aberration
            fan = new Goptical::Analysis::RayFan(*eye.sys);
            
            eye.sys->enable_single<Sys::Source>(*eye.source_point);
            ref<Data::Plot> abber_plot = fan->get_plot(Goptical::Analysis::RayFan::EntranceHeight,
                                                Goptical::Analysis::RayFan::LongitudinalDistance);
            
            renderer.set_page(1);
            abber_plot->draw(renderer);

            eye.Diopters(true);
            break;
        }
        case 2:
        {
            int _diop = 0;
            Goptical::Io::RendererSvg renderer("img/eye.svg", 1200,800);
            renderer.set_margin_ratio(0.1, 0.1, 0.1, 0.1);
            renderer.set_page_layout(1,4);
            for (int i = 0; i < 4; i++)
            {    
                Eye eye = Eye(); 
                eye.set_params(model);
                eye.SchematicEye();
                eye.EyeTracer(object_distance, off_axis);
                //EyePlots();
                
                renderer.set_page(i);
                
                eye.sys->enable_single<Sys::Source>(*eye.source_rays);
                eye.tracer->get_trace_result().set_generated_save_state(*eye.source_rays);
                eye.sys->draw_2d_fit(renderer);
                eye.sys->draw_2d(renderer);                
                eye.tracer->trace();
                eye.tracer->get_trace_result().draw_2d(renderer);

                //delete &eye;
                _diop += 2;
                
            }
        }    
        
    }
}


void Analysis::LSAanalysis(float object_distance, float off_axis,
    std::string model)
{
    std::ofstream outputfile;
    outputfile.open ("dat/EyeLSA.csv", std::ios::trunc);
    float AGE, LensAccomm, PupilSize, AccommOptPower, 
            RelaxedOptPower, Defocus_diopters;

    for (int i = 0; i < 20; i++)
    {   
        for (int j = 1; j < 21; j++)
        {
            for (int k = 10; k <25; k++)
            {
                LensAccomm = i / 2.0;
                PupilSize = j / 4.0;
                AGE = k;
                
                Eye eye = Eye(); 
                eye.set_params( LensAccomm, PupilSize, model, AGE );
                eye.SchematicEye();
                eye.EyeTracer(object_distance, off_axis);
                
                AccommOptPower = eye.FindOpticalPower(1);
                RelaxedOptPower = eye.FindOpticalPower(2);
                Defocus_diopters = eye.Diopters(true);
                
                outputfile << LensAccomm << "," << PupilSize << "," << AGE << "," << 
                            AccommOptPower << "," << RelaxedOptPower << "," << 
                            Defocus_diopters << std::endl;

            }
        }
    }
    
    outputfile.close();
}


void Analysis::SpotPlot(float object_distance=100000, 
                    float off_axis=5, std::string model="dubbleman", 
                    float age=20, float pupil_size=3, float diopters=0)
{
    Eye eye = Eye(); 
    //float diopters, pupil_size, age;
    //diopters = _get_input("lens diopters (mm)");
    //pupil_size = _get_input("pupil size (mm)");
    //age = _get_input("age (years)");

    eye.set_params(diopters, pupil_size, model, age);
    eye.SchematicEye();
    eye.EyeTracer(object_distance, off_axis);

    eye.sys->enable_single<Sys::Source>(*eye.source_point);
    Goptical::Io::RendererSvg     renderer("img/spot.svg", 600,600, 
        Io::rgb_black);
    //renderer.set_margin_ratio(0.1, 0.1, 0.1, 0.1);
    
    eye.sys->get_tracer_params().set_default_distribution(
                            Trace::Distribution(Trace::HexaPolarDist, 100)); 
    //renderer.set_page_layout(1, 1);

    Goptical::Analysis::Spot spot(*eye.sys);
    
    //renderer.set_page(0);
    spot.draw_diagram(renderer);

}

void Analysis::EyePlots(float object_distance=100000, float off_axis=5, 
                    std::string model="navarro", float age=20, float pupil_size=3, 
                    float diopters=0)
{ 
    Eye eye = Eye(); 
    Goptical::Io::RendererSvg renderer("img/eye_analysis.svg", 300,1200, Io::rgb_black);

    renderer.set_margin_ratio(0.1, 0.1, 0.1, 0.1);
    renderer.set_page_layout(1,4);

    int _diop = 0;
    for (int i = 0; i < 4; i++)
    {    
        
        eye.set_params(_diop, eye.pupil_size, eye.model, eye.age);
        eye.SchematicEye();
        eye.EyeTracer(object_distance, off_axis);
        
        renderer.set_page(0);
        eye.sys->get_tracer_params().set_default_distribution(
                            Trace::Distribution(Trace::HexaPolarDist, 2550)); 
        
        
        Goptical::Analysis::Spot spot(*eye.sys);
        
        spot.draw_diagram(renderer);    

        ref<Data::Plot> plot = spot.get_encircled_intensity_plot(100);

        plot->draw(renderer);                

        renderer.set_page(1);
        renderer.set_page(2);
        
        //Goptical::Analysis::Spot spot(*eye.sys);
        Goptical::Analysis::Focus focus(*eye.sys);
        eye.image->set_plane(focus.get_best_focus());

        spot.draw_diagram(renderer);    

        //ref<Data::Plot> plot = spot.get_encircled_intensity_plot(100);

        renderer.set_page(3);
        plot->draw(renderer);                
    }    
}


arma::vec Analysis::PSF(arma::vec intensity, arma::vec xvals, bool symmetric=false)
{
    float samples;
    samples = intensity.n_elem;
    arma::vec psf = arma::zeros<arma::vec>(1, samples);
    arma::vec psftotal = arma::zeros<arma::vec>(1, samples * 2);

    // we have an integral, therefore take the deriv to get rays / bin
    arma::vec deriv = arma::zeros<arma::vec>(samples);

    deriv(0) = intensity(0);
    deriv.rows(1, samples) = intensity.rows(1, samples) - intensity.rows(0, samples - 1);

    float radius0, radius1, area;
    for (int i = 0; i < samples - 1; i++)
    {
        // account for increasing size of area
        radius0 = xvals(i);
        radius1 = xvals(i + 1);
        
        // subtract inner and outer circle area to get sliver of interest
        area = (PI * pow(radius1, 2)) - (PI * pow(radius0, 2));

        // deriv = amount in each circle; then divide by area
        psf(i) = deriv(i) / area;
    } 

    // normalize so that PSF has integral of 1.
    psf = psf / arma::sum(psf);

    psftotal.rows(1, samples + 1) = arma::fliplr(psf);
    psftotal.rows(0, samples) = psf.rows(1, samples - 1);

    if (symmetric) { return psftotal; }
    else { return psf; }
}

arma::vec Analysis::MTF(arma::vec psf)
{
    arma::vec temp;

    arma::vec mtf = arma::zeros<arma::vec>(psf.n_elem);

    temp = arma::abs(arma::fft(psf));
    temp = arma::real(temp);

    // normalize MTF
    mtf = temp / psf.max();
    return mtf;
}
}

File: python/eye_analysis.py
#! /usr/bin/env python
from __future__ import division
import numpy as np
import matplotlib.pylab as plt
import sys

from base import optics as o
from base import data as dm


def loadData(downsample=1.0):
    '''
    '''
    dat = np.genfromtxt('dat/Intensity.csv', delimiter = ',', 
        dtype="f8,f8,f8,f8,f8,f8,f8,S20,f8,f8",
        names=True, skip_header=0)

    intensity = {}
    meta = {}

    meta['model'] = dat['model'][0]
    meta['iterations'] = int(dat['iterations'][0])
    meta['samples'] = int(dat.shape[0] / meta['iterations'])

    # Preallocate memory:
    intensity['rawintensity'] = np.zeros((meta['iterations'],
                                    meta['samples'])) 
    intensity['intensity'] = np.zeros((meta['iterations'], 
                                    meta['samples']))

    meta['lens_accom'] = np.zeros(meta['iterations'])
    meta['pupil_size_%'] = np.zeros(meta['iterations'])
    meta['off_axis_*'] = np.zeros(meta['iterations'])
    meta['object_distance'] = np.zeros(meta['iterations'])
    meta['age_&'] = np.zeros(meta['iterations'])
    meta['eye_length_%'] = np.zeros(meta['iterations'])

    for i in range(0, meta['iterations']):
        ind1 = i * meta['samples']
        ind2 = (i + 1) * meta['samples']

        intensity['rawintensity'][i,:] = dat['encircled_intensity'][ind1:ind2]
        intensity['intensity'][i,:] = (dat['encircled_intensity'][ind1:ind2]
                             / np.max(intensity['rawintensity'][i,:]))

        meta['lens_accom'][i] = dat['lens_focus_D'][ind1]
        meta['pupil_size_%'][i] = dat['pupil_size_mm'][ind1]
        meta['off_axis_*'][i] = dat['offaxis_deg'][ind1]
        meta['object_distance'][i] = dat['obj_distance_mm'][ind1]
        meta['age_&'][i] = dat['age_y'][ind1]
        meta['eye_length_%'][i] = dat['axial_len_mm'][ind1]

    # set up parameters of the eye:
    xvals = dat['radius_mm'][0: meta['samples']]

    # retinal image, here is radius of image area
    meta['retImg'] = np.max(xvals) # 1/2 size of image in mm 
    radians = 2 * np.arctan(meta['retImg'] / (meta['eye_length_%'][0]))
    meta['deg'] = dm.rad2deg(radians)
    meta['mm/deg'] = meta['retImg'] * 2 / meta['deg']
    meta['xvals'] = xvals

    variable = findDependentVar(meta)
    printParams(meta)

    return intensity, meta


def findDependentVar(meta):
    '''
    '''
    variable = None

    for key in meta.iterkeys():
        try :
            if len(meta[key]) > 0:
                if key != "model":
                    if meta[key][0] != meta[key][1]:
                        variable = key
        except (IndexError, TypeError):
            pass
    return variable


def printParams(meta):
    '''
    '''
    print ' '
    for key in meta.iterkeys():
        if key != 'xvals':
            out = (key.replace('_',' ')
                .replace('%','(mm)')
                .replace('&','(y)')
                .replace('D', '(D)')
                .replace('*', '$\\degree$'))
            if len(out) < 3:
                print out, "\t\t\t\t||\t", meta[key]
            elif len(out) < 7:
                print out, "\t\t\t||\t", meta[key]
            elif len(out) < 15:
                print out, "\t\t||\t", meta[key]
            else:
                print out, " \t||\t", meta[key]
    print ' '


def findPlottingData(data, arg1, type1, arg2, 
                type2, arg3=None, type3=None):
    '''Called by LSA and power plots
    '''

    ind1 = data[type1] == arg1
    ind2 = data[type2] == arg2
    
    if arg3 == None:
        loc = ind1 & ind2 == True
    else:
        ind3 = data[type3] == arg3
        loc = ind1 & ind2 & ind3 == True

    return loc


def loadLSA_data():
    '''
    '''
    dat = np.genfromtxt('../dat/EyeLSA.csv', delimiter = ',')
    
    lsa_data = {
                'lens_accommodation_D': dat[:, 0], 
                'pupil_size_mm': dat[:, 1], 
                'age': dat[:, 2], 
                'accomPower': dat[:, 3], 
                'relaxPower': dat[:, 4],
                'defocus': dat[:, 5]
                }
    return lsa_data


def getMTF(meta, intensity):
    '''
    '''
    if 'PSF' not in intensity:
        intensity = getPSF(meta, intensity)

    intensity['MTF'] = np.zeros((meta['iterations'], 
                                    meta['samples'] ))
    for i in range(0, meta['iterations']):
        # normalize MTF
        intensity['MTF'][i, :] = o.MTF(intensity['PSFtotal'][i, :])

    return intensity


def getPSF(meta, intensity):
    '''
    '''
    intensity['PSF'] = np.zeros((meta['iterations'],
                                meta['samples']))

    intensity['PSFtotal'] = np.zeros((meta['iterations'],
                                 (meta['samples'] * 2) + 1))

    for i in range(0, meta['iterations']):
        intensity['PSF'][i, :], intensity['PSFtotal'][i, :] = (
            o.PSF(intensity['rawintensity'][i, :], meta['samples'])
        )
    return intensity
